### PR TITLE
Resolves #37695 - Use clsx instead of classnames

### LINKED ISSUE #37695: Monorepo tooling: replace classnames with clsx
**Let's aim to replace `classnames` with `clsx` across the monorepo.**

`clsx` is now used in Calypso (see https://github.com/Automattic/wp-calypso/pull/91408), in Gutenberg (see https://github.com/WordPress/gutenberg/pull/61138), and in `@wordpress` packages (related: https://github.com/Automattic/jetpack/pull/37669). 

With that mind, it makes sense for Jetpack to make the switch as well, for better consistency.

******

Internal reference: p4TIVU-aYc-p2

### PR DESCRIPTION
**Pull request by tbradsha**
Resolves #37695 - Use clsx instead of classnames

## Proposed changes:

`clsx` is [faster](https://github.com/lukeed/clsx/tree/925494cf31bcd97d3337aacd34e659e80cae7fe2/bench) and smaller  than `classnames`. Also, it is now used in Calypso (Automattic/wp-calypso#91408), in Gutenberg (WordPress/gutenberg#61138), and in `@wordpress` packages.

For consistency, this PR matches those changes:

* a rule was added to prevent future use of `classnames`: 6b5518d1653f1fcf4664b852c2b1cf6cad89cfa5
* `classnames` was replaced with `clsx`: b45ef169f05041337f0a52773237b3eb428febc0
* `classnames/dedupe` does not have an equivalent (lukeed/clsx#13), so its one use was refactored using `@wordpress/token-list`: a060f549765f07d892e4548165f8f650d6e614dd

Caveats:
* `jetpack-mu-wpcom` has [its own implementation](https://github.com/Automattic/jetpack/blob/fix/37695/classnames_to_clsx/projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/src/utils.ts#L9) of `classNames`; I've left that intact for now. If desired, we can take a look in a separate PR.
* `@automattic/social-previews` is a dependency, and the latest published version ([`2.0.1-beta.13`](https://www.npmjs.com/package/@automattic/social-previews)) still has a `classnames` dependency. This was removed in the aforementioned Calypso PR, but as of yet has not been published.

Internal reference: p4TIVU-aYc-p2

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

Check the code for any typos or mistakes. Pay particularly close attention to the dedupe refactor.

### CODE REVIEW

**Comment by jeherve** [comment]
> jetpack-mu-wpcom has [its own implementation](https://github.com/Automattic/jetpack/blob/fix/37695/classnames_to_clsx/projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/src/utils.ts#L9) of classNames; I've left that intact for now. If desired, we can take a look in a separate PR.

@arcangelini I see that was added as part of the initial commit of Verbum Comments, in its original repo. Was there a specific reason you went with a custom implementation? Could you take a look at moving it to `clsx`?

File: projects/plugins/jetpack/extensions/blocks/videopress/deprecated/v2/utils.js [hunk 1]
```diff
@@ -25,31 +24,29 @@ export const ASPECT_RATIOS = [
...
-			return accumulator;
-		},
-		{ 'wp-has-aspect-ratio': false }
-	);
 
-	const customClasses = classnames( className, {
+	const aspectRatioClassNames = ASPECT_RATIOS.reduce( ( accumulator, ratio ) => {
+		accumulator.push( ratio.className );
+		return accumulator;
+	}, [] );
```
Comment (anomiex) [inline projects/plugins/jetpack/extensions/blocks/videopress/deprecated/v2/utils.js]: If you want you could simplify this a bit
```suggestion
	const aspectRatioClassNames = ASPECT_RATIOS.map( ratio => ratio.className );
```

File: projects/plugins/jetpack/extensions/blocks/videopress/deprecated/v2/utils.js [hunk 2]
```diff
@@ -25,31 +24,29 @@ export const ASPECT_RATIOS = [
...
-	const customClassList = new TokenList( customClasses );
-	let videoPressClasses = classnames( className, videoPressClassNames );
+	const videoPressClassTokenList = new TokenList( className );
+	videoPressClassTokenList.add( videoPressClassNames );
 
-	customClassList.forEach( customClass => {
-		videoPressClasses = videoPressClasses.replace( customClass, '' );
+	customClassTokenList.forEach( customClass => {
+		videoPressClassTokenList.remove( customClass );
 	} );
```
Comment (anomiex) [inline projects/plugins/jetpack/extensions/blocks/videopress/deprecated/v2/utils.js]: Looks like you could simplify this too, `...` works on a TokenList.
```suggestion
	videoPressClassTokenList.remove( ...customClassTokenList );
```

**Review by anomiex — COMMENTED** [review]
Looks reasonable to me. A few suggestions for simplifying the dedupe replacement inline.

File: projects/plugins/jetpack/extensions/blocks/videopress/deprecated/v2/utils.js (hunk 1 above)
Comment (tbradsha) [reply projects/plugins/jetpack/extensions/blocks/videopress/deprecated/v2/utils.js]: Ha. Clearly I wasn't thinking.

File: projects/plugins/jetpack/extensions/blocks/videopress/deprecated/v2/utils.js (hunk 2 above)
Comment (tbradsha) [reply projects/plugins/jetpack/extensions/blocks/videopress/deprecated/v2/utils.js]: Nice. I did experiment with passing an array of items and passing the class string, but your version works splendidly.

**Review by anomiex — COMMENTED** [review]
LGTM. Not hitting "Approve" just yet to give Agora a chance to review.

**Comment by DaniGuardiola** [comment]
Just a quick drive-by comment, I was the one who originally started this transition and was personally responsible for the Gutenberg PR. I managed to replace the dedupe usages without much complex logic or the new dependency you're using here. The code affected seems very similar, probably copy-pasted. I suggest you take a look at what I did there in case it helps you simplify this PR and achieve better consistency.

Note that I don't remember the exact approach I took, but I might have decided to simply not dedupe at all in some cases, since it doesn't really affect any functionality and the trade-off of doing the computation is probably not worth it in the first place.

Just a suggestion :) thanks for applying this transition here <3

**Comment by anomiex** [comment]
> but I might have decided to simply not dedupe at all, since it doesn't really affect any functionality

You may want to double-check that.
```
$ node -e 'const classnames = require( "classnames" ); console.log( classnames( "aa bb cc", { bb: false } ) );'
aa bb cc
$ node -e 'const clsx = require( "clsx" ); console.log( clsx( "aa bb cc", { bb: false } ) );'
aa bb cc
$ node -e 'const classnames = require( "classnames/dedupe" ); console.log( classnames( "aa bb cc", { bb: false } ) );'
aa cc
```

**Comment by DaniGuardiola** [comment]
Ah, yep I replaced that behavior with a very very simple alternative. If you dig in my PR you'll find my approach. With that sentence in my comment I meant cases where the resulting class name is something like "class class".

**Comment by anomiex** [comment]
Glancing at your PR, the following seem likely to be broken to me
* https://github.com/WordPress/gutenberg/blob/200f7c7a210c35eb9c004da9a3e4f1062f471d57/packages/block-library/src/cover/edit.native.js#L301-L303
* https://github.com/WordPress/gutenberg/blob/200f7c7a210c35eb9c004da9a3e4f1062f471d57/packages/block-library/src/embed/embed-preview.js#L119-L121

Also, in https://github.com/WordPress/gutenberg/blob/200f7c7a210c35eb9c004da9a3e4f1062f471d57/packages/block-library/src/embed/util.js#L191-L193 you might mangle things if a class in `outputClassNames` has one of the classes in `aspectRatioClassNames` as a substring. For example, pass in "foo wp-has-aspect-ratio-xxx bar", you'll get back "foo -xxx bar".

**Comment by tbradsha** [comment]
>  I managed to replace the dedupe usages without much complex logic or the new dependency you're using here.

The one `dedupe` instance here required negation (removal of various classes). We don't use `clsx` or `classnames` for it. I'm not sure what new dependency you're referring to?

> The code affected seems very similar, probably copy-pasted. I suggest you take a look at what I did there in case it helps you simplify this PR and achieve better consistency.

I can't speak to the origin of the code, but in this case it was an independent refactor; looking now at that other PR it seems the goal is slightly different, but luckily it isn't too complex!

**Review by renatoagds — DISMISSED** [review]
Just got an eyes on VideoPress and AI stuffs. LGTM

**Review by zinigor — APPROVED** [review]

**Comment by tbradsha** [comment]
Internal announcement: pdWQjU-Mt-p2

**Pull request merged**